### ISSUE TITLE
[BUG] `frame.evaluate` is flaky

### ISSUE THREAD

**Issue by mhils** [microsoft/playwright#8724]
Hi folks! I'm trying to track down a race condition where `frame.evaluate` is not executing. Here's a minimal example:

```javascript
const {chromium} = require('playwright');

(async () => {
    const browser = await chromium.launch({headless: false});
    const page = await browser.newPage();
    page.on("console", (msg) => msg.type() === "info" && console.info(msg.text()));

    await page.goto("https://www.westfalen-blatt.de/");
    await page.waitForTimeout(5_000);
    console.log("".padStart(120, "#"))

    for (let frame of page.frames()) {
        if (frame.url().startsWith("https://cdn.privacy-mgmt.com")) {
            console.log(">>", frame.url());
            await frame.evaluate(`console.info(" <<", location.href)`);
        }
    }

    console.log("all done");
    await browser.close();
})();

```

The expected output would be this:
```
########################################################################################################################
>> https://cdn.privacy-mgmt.com/index.html?...
 << https://cdn.privacy-mgmt.com/index.html?...
all done
```
Unfortunately I seem to be hitting a race condition with >10% probability where the output stops after the `>>` line. I can confirm visually that the iframe remains in the document at all times.

I've recorded a log with `$env:DEBUG="pw:protocol"`: https://gist.github.com/mhils/e7cba45267c01036b1c305f64acb605c. Here's an excerpt from the end:

```
2021-09-06T17:14:10.9081127: pw:protocol ◀ RECV {"method":"Page.lifecycleEvent","params":{"frameId":"83C7EC0D142F5B666D599AD9B0775B48","loaderId":"63E9CA5FCBD93E182096E6030C678449","name":"InteractiveTime","timestamp":27155.624692},"sessionId":"0F4B3E1E274E92E173148D7E58E872BA"}
2021-09-06T17:14:11.0589127: ########################################################################################################################
2021-09-06T17:14:11.0589127: >> https://cdn.privacy-mgmt.com/index.html?message_id=544167&consentUUID=5e789f8c-2248-45e0-88c3-ed6ac9195aa9&requestUUID=88ad162c-0b44-4bca-9816-620d3e7069d5&preload_message=true @ frame@db64aecf5e38249e226daf7cf7a8d2ba
2021-09-06T17:14:16.5958191: pw:protocol ◀ RECV {"method":"Runtime.executionContextDestroyed","params":{"executionContextId":9},"sessionId":"0F4B3E1E274E92E173148D7E58E872BA"}
```

Looking at the timestamps, `frame.evaluate` does not seem to trigger any protocol message. Do you folks have any idea what could be causing this?

**System Information:**
```
 - Playwright: 1.15.0-next-alpha-sep-6-2021
 - OS: Windows 10 10.0.19043
 - Node: 16.8.0
```

**Comment by pavelfeldman** [comment]
I can't reproduce this one. Tried a bunch of times, headless and headed, different timeout values.

**Comment by mhils** [comment]
Yeah, this might be a bit of a tricky one. I also failed to reproduce this on a Linux VM I planned to share with you just now. 
It still does happen locally on my Windows machine. Can I provide you with anything helpful beyond the `pw:protocol` log above?

The only thing that looks _somewhat_ suspicious to my untrained eye is the `Page.frameDetached` event at `2021-09-06T17:14:05.3309459`, which references the target frame. Total shot in the dark though, I'm not too familar with the devtools API.

**Comment by pavelfeldman** [comment]
I'll have to boot windows and try to repro - parsing the protocol on something flaky isn't fun...

**Comment by dgozman** [comment]
Thank you for the logs, I was able to get a working theory out of them 😄  This seems like a problem with out-of-process iframes.

I've landed a fix in #8926 that will hopefully fix the issue, and so I will tentatively close this issue. Please file a new one if the issue persists after the next v1.15 release.

**Comment by mhils** [comment]
@dgozman: Spot on! 🎯 

I can't reproduce the original issue anymore since upgrading from `1.15.0-next-alpha-sep-14-2021` to `1.15.0-next-alpha-sep-16-2021`, thank you for taking care of this! 🍰 😃
